File: src/ISA/ISA/ArcTypes/CompositeColumn.fs
namespace ARCtrl.ISA

open Fable.Core

[<AttachMembers>]
type CompositeColumn = {
    Header: CompositeHeader
    Cells: CompositeCell []
} with
    static member create (header: CompositeHeader, ?cells: CompositeCell []) = 
        let cells = Option.defaultValue [||] cells
        {
            Header = header
            Cells = cells
        }

    /// Returns true if header and cells are a valid combination. E.g. Term header with term or unitized cells. IO header with freetext cells.
    ///
    /// ?raiseExeption: Default false. Set true if this function should raise an exception instead of return false.
    // TODO! Do not only check cells.Head
    member this.Validate(?raiseException: bool) =
        let raiseExeption = Option.defaultValue false raiseException
        let header = this.Header
        let cells = this.Cells
        match header, cells with
        // no cell values will be handled later and is no error case
        | _, emptyCell when cells.Length = 0 -> 
            true
        | isTerm when header.IsTermColumn && (cells.[0].isTerm || cells.[0].isUnitized) -> 
            true
        | isNotTerm when not header.IsTermColumn && cells.[0].isFreeText -> 
            true
        | h, c -> 
            if raiseExeption then 
                let exampleCells = c.[0]
                let msg = $"Invalid combination of header `{h}` and cells `{exampleCells}`"
                failwith msg
            // Maybe still return `msg` somehow if `raiseExeption` is false?
            false

    /// <summary>
    /// Returns an array of all units found in the cells of this column. Returns None if no units are found.
    /// </summary>
    member this.TryGetColumnUnits() =
        let arr = [|
            for cell in this.Cells do
                if cell.isUnitized then
                    let _, unit = cell.AsUnitized
                    unit
        |]
        if Array.isEmpty arr then None else Some arr

File: js/ISA/ISA/ArcTypes/CompositeColumn.js
import { defaultArg } from "../../../fable_modules/fable-library.4.1.4/Option.js";
import { min } from "../../../fable_modules/fable-library.4.1.4/Double.js";
import { Record } from "../../../fable_modules/fable-library.4.1.4/Types.js";
import { CompositeHeader_$reflection } from "./CompositeHeader.js";
import { CompositeCell_$reflection } from "./CompositeCell.js";
import { record_type, array_type } from "../../../fable_modules/fable-library.4.1.4/Reflection.js";

export class CompositeColumn extends Record {
    constructor(Header, Cells) {
        super();
        this.Header = Header;
        this.Cells = Cells;
    }
    static create(header, cells) {
        return new CompositeColumn(header, defaultArg(cells, []));
    }
    validate(raiseException) {
        const this$ = this;
        const raiseExeption = defaultArg(raiseException, false);
        const header = this$.Header;
        const cells = this$.Cells;
        if (cells.length === 0) {
            return true;
        }
        else if (header.IsTermColumn && (cells[0].isTerm ? true : cells[0].isUnitized)) {
            return true;
        }
        else if (!header.IsTermColumn && cells[0].isFreeText) {
            return true;
        }
        else {
            const c = cells;
            if (raiseExeption) {
                throw new Error(`Invalid combination of header \`${header}\` and cells \`${c[min(c.length, 3)]}\``);
            }
            return false;
        }
    }
}

export function CompositeColumn_$reflection() {
    return record_type("ISA.CompositeColumn", [], CompositeColumn, () => [["Header", CompositeHeader_$reflection()], ["Cells", array_type(CompositeCell_$reflection())]]);
}

